hubblestack/loader.py

```diff
@@ -0,0 +1,1292 @@
+                        log.warning(
+                            '%s.__virtual__() is wrongly returning `None`. '
+                            'It should either return `True`, `False` or a new '
+                            'name. If you\'re the developer of the module '
+                            '\'%s\', please fix this.', mod.__name__, module_name
+                        )
+
+                    return (False, module_name, error_reason, virtual_aliases)
+
+                # At this point, __virtual__ did not return a
+                # boolean value, let's check for deprecated usage
+                # or module renames
+                if virtual is not True and module_name != virtual:
+                    # The module is renaming itself. Updating the module name
+                    # with the new name
+                    log.trace('Loaded %s as virtual %s', module_name, virtual)
+
+                    if not hasattr(mod, '__virtualname__'):
+                        salt.utils.versions.warn_until(
+                            'Hydrogen',
+                            'The \'{0}\' module is renaming itself in its '
+                            '__virtual__() function ({1} => {2}). Please '
+                            'set it\'s virtual name as the '
+                            '\'__virtualname__\' module attribute. '
+                            'Example: "__virtualname__ = \'{2}\'"'.format(
+                                mod.__name__,
+                                module_name,
+                                virtual
+                            )
+                        )
+
+                    if virtualname != virtual:
+                        # The __virtualname__ attribute does not match what's
+                        # being returned by the __virtual__() function. This
+                        # should be considered an error.
+                        log.error(
+                            'The module \'%s\' is showing some bad usage. Its '
+                            '__virtualname__ attribute is set to \'%s\' yet the '
+                            '__virtual__() function is returning \'%s\'. These '
+                            'values should match!',
+                            mod.__name__, virtualname, virtual
+                        )
+
+                    module_name = virtualname
+
+                # If the __virtual__ function returns True and __virtualname__
+                # is set then use it
+                elif virtual is True and virtualname != module_name:
+                    if virtualname is not True:
+                        module_name = virtualname
+
+        except KeyError:
+            # Key errors come out of the virtual function when passing
+            # in incomplete grains sets, these can be safely ignored
+            # and logged to debug, still, it includes the traceback to
+            # help debugging.
+            log.debug('KeyError when loading %s', module_name, exc_info=True)
```

Comment:
Can we make it ERROR level. I faced this issue when I tried to load some grains, there was no error and my grain was not loaded.
Making it error can upfront convey the issues in grains. Anyway, we don't put incomplete grains in our folders.

Reply:
Yes. This is a good idea, but ... we'll have to keep an eye on this. I imagine it'll get very noisy and send millions of boring duplicate errors to Splunk.